### DISCUSSION TITLE
Usage/Experiences when using Cosmic on OpenRC

### DISCUSSION THREAD

**Discussion by simonvanderveldt** [fsvm88/cosmic-overlay#30] [General]
I just wanted to create a place to share experiences/suggestions/fixes/etc for running Cosmic on a system that's running OpenRC. Thanks to the fixes in #11 emerging/compilation is working fine and Cosmic starts/runs as well :tada: 

But things like pipewire, notifications, power settings don't work unless a service is running for them. I was wondering how people are handling this? Autostart isn't supported yet on Cosmic (https://github.com/pop-os/cosmic-session/issues/67) and OpenRC doesn't have support for user-services yet (https://github.com/OpenRC/openrc/pull/723).
I assume Cosmic on systemd uses user-services for all these components?

P.S. A discussion instead of a issue would've probably been better for this, but since discussions have been disabled hope this is OK for you :)

**Comment by fsvm88** [comment]
Have enabled discussions =) It's a feature I never used, and it was disabled by default with project creation =)

Feel free to move there.

Regarding user services: this is mostly automatically handled by systemd and/or Cosmic, I'm not doing anything special with it. IIRC cosmic-settings-daemon and cosmic-notifications are spawned automatically by Cosmic after first login (not sure which component, I'd guess cosmic-session).

I recall with the transition to pipewire I had to manually enable some services for the user, but that was a year ago now, I'm not sure what's the default state on a newly-created user. See [Gentoo Wiki page](https://wiki.gentoo.org/wiki/PipeWire). 

Although I ran with it for a good decade, I am not using OpenRC anymore because pretty much every DE is heavily relying on systemd nowadays, and I really like not having to think about service restarts on servers.

**Comment by simonvanderveldt** [comment]
> Have enabled discussions =) It's a feature I never used, and it was disabled by default with project creation =)
> 
> Feel free to move there.

Thanks! :+1:  It's just semantics I guess in the end, issue vs discussion. I don't think I can move it, AFAIK only repo owners/with relevant repo permissions can do so.


> Regarding user services: this is mostly automatically handled by systemd and/or Cosmic, I'm not doing anything special with it. IIRC cosmic-settings-daemon and cosmic-notifications are spawned automatically by Cosmic after first login (not sure which component, I'd guess cosmic-session).

Thanks for the pointer, I thought that this was all being handled by `xdg-desktop-autostart.target` but that doesn't seem to be the case. Booted the Cosmic alpha 3 ISO and this is the process tree
![screenshot-2024-11-10-14-37-03](https://github.com/user-attachments/assets/05ebd2cf-a5e1-4df1-a308-393d5b1fc50a)
So `greetd` starts `cosmic-session`, which starts pretty much all cosmic components including `cosmic-notifcations`.
Interestingly enough `cosmic-notifcations` is missing for me on my gentoo system
```
root      3227     1  0 12:07 ?        00:00:00 /usr/sbin/gdm
root      4188  3227  0 12:08 ?        00:00:00  \_ gdm-session-worker [pam/gdm-password]
simon     4212  4188  0 12:08 tty8     00:00:00      \_ /usr/libexec/gdm-wayland-session --register-session /us
simon     4215  4212  0 12:08 tty8     00:00:00          \_ dbus-daemon --print-address 3 --session
simon     4217  4212  0 12:08 tty8     00:00:00          \_ /usr/bin/cosmic-session
simon     4225  4217 52 12:08 tty8     01:49:45              \_ cosmic-comp
simon     4279  4225  0 12:08 tty8     00:01:36              |   \_ Xwayland :0 -verbose -rootless -terminate -
simon     4305  4217  0 12:08 tty8     00:00:00              \_ cosmic-settings-daemon
simon     4314  4305  0 12:08 tty8     00:00:00              |   \_ /usr/libexec/geoclue-2.0/demos/agent
simon     4312  4217  0 12:08 tty8     00:00:51              \_ cosmic-panel
simon     4652  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-status-area
simon     4654  4312  0 12:08 tty8     00:00:03              |   \_ cosmic-applet-time
simon     4655  4312  0 12:08 tty8     00:00:38              |   \_ cosmic-applet-audio
simon     4659  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-power
simon     4664  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-battery
simon     4674  4312  0 12:08 tty8     00:00:03              |   \_ cosmic-applet-workspaces
simon     4676  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-tiling
simon     4678  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-panel-button com.system76.CosmicWork
simon     4680  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-bluetooth
simon     4682  4312  0 12:08 tty8     00:00:00              |   \_ cosmic-applet-network
simon     4683  4312  0 12:08 tty8     00:00:04              |   \_ cosmic-app-list
simon     4313  4217  0 12:08 tty8     00:00:00              \_ cosmic-app-library
simon     4316  4217  0 12:08 tty8     00:00:22              \_ cosmic-launcher
simon    25512  4316  0 15:36 tty8     00:00:00              |   \_ pop-launcher
simon    25521 25512  0 15:36 tty8     00:00:00              |       \_ /usr/lib/pop-launcher/plugins/cosmic_to
simon    25522 25512  0 15:36 tty8     00:00:00              |       \_ /usr/lib/pop-launcher/plugins/pop_shell
simon     4325  4217  0 12:08 tty8     00:00:09              \_ cosmic-workspaces
simon     4332  4217  0 12:08 tty8     00:00:18              \_ cosmic-osd
simon     4340  4217  0 12:08 tty8     00:00:01              \_ cosmic-bg
simon     4348  4217  0 12:08 tty8     00:00:07              \_ cosmic-greeter
simon     4359  4217  0 12:08 tty8     00:00:02              \_ cosmic-files-applet
simon     4368  4217  0 12:08 tty8     00:00:00              \_ cosmic-idle
simon     4377  4217  0 12:08 tty8     00:00:15              \_ /usr/libexec/xdg-desktop-portal-cosmic
```
(it's still using GDM since I'm coming from GNOME, but I doubt that's the reason it isn't working).

I'll see if I can figure out why `cosmic-notifications` isn't being started and create a PR if there's anything to fix in any of the ebuilds.

**Comment by fsvm88** [comment]
Managed to convert it! :-D

Another (hopefully useful) pointer: check your logs for errors. In the past (up to alpha2) these were sometimes difficult to see, and also to distinguish between "error reading file (e.g: local config) but I can continue" and "fatal error I'm aborting now". This is the reason why this issue #8 exists: cosmic-settings-daemon would compile fine without geoclue, but it would crash at startup, and would keep failing any time you changed a setting (leading to inability to persist any changed settings). There was absolutely no UI indication something went wrong, just the side-effect that no settings would stick.

**Comment by simonvanderveldt** [comment]
Thanks! I was already looking at the logs, haven't really been able to find any real cause or fix so far, so created an issue in the main cosmic repo https://github.com/pop-os/cosmic-epoch/issues/1237, hope they have some suggestions there to troubleshoot it further.

**Comment by vadorovsky** [comment]
I can confirm the same issue with cosmic-notifications.

@simonvanderveldt Does the XDG autostart work when you're starting COSMIC with GDM? I have to launch `/usr/bin/gentoo-pipewire-launcher restart &` and `/usr/libexec/xdg-desktop-portal --replace` manually in terminal when I'm using cosmic-greeter. I think I want to avoid usage of GDM and I'm actually thinking about making a PR with XDG autostart support in cosmic-session.

Another issue I'm facing is that I cannot run Steam (via flatpak). I'm seing the following error:

```
The unofficial Steam Flatpak app requires a correctly-configured desktop
session, which must provide the DISPLAY environment variable to the
D-Bus session bus activation environment.

On systems that use systemd --user, the DISPLAY environment variable must
also be present in the systemd --user activation environment.

This is usually achieved by running:

    dbus-update-activation-environment DISPLAY

during desktop environment startup.

For more details, please see:
https://github.com/ValveSoftware/steam-for-linux/issues/10554
```

Even though the dbus service is up and running. I guess I'm missing yet another piece which is supposed to be autostarted.

For reference, here is my process tree:

```
root      2828     1  0 Nov16 ?        00:00:00 /usr/bin/greetd
root      3051  2828  0 Nov16 tty1     00:00:00  \_ /usr/bin/greetd --session-worker 12
vadorov+  3251  3051  0 Nov16 tty1     00:00:00      \_ -/usr/bin/nu -c /usr/bin/start-cosmic --in-login-shell
vadorov+  3270  3251  0 Nov16 tty1     00:00:00          \_ /usr/bin/dbus-run-session -- /usr/bin/cosmic-session
vadorov+  3273  3270  0 Nov16 tty1     00:00:00              \_ /usr/bin/cosmic-session
vadorov+  3451  3273  0 Nov16 tty1     00:14:42              |   \_ cosmic-greeter
vadorov+  3452  3273  0 Nov16 tty1     00:00:11              |   \_ cosmic-files-applet
vadorov+  3342  3273  5 Nov16 tty1     01:23:34              |   \_ cosmic-osd
vadorov+  3338  3273  0 Nov16 tty1     00:00:00              |   \_ cosmic-launcher
vadorov+  3616  3338  0 Nov16 tty1     00:00:00              |   |   \_ pop-launcher
vadorov+  3630  3616  0 Nov16 tty1     00:00:00              |   |       \_ /usr/lib/pop-launcher/plugins/pop_shell/pop-shell
vadorov+  3329  3273  0 Nov16 tty1     00:00:00              |   \_ cosmic-notifications
vadorov+  3339  3273  0 Nov16 tty1     00:00:05              |   \_ cosmic-workspaces
vadorov+  3351  3273  0 Nov16 tty1     00:00:05              |   \_ cosmic-bg
vadorov+  3332  3273  0 Nov16 tty1     00:04:53              |   \_ cosmic-panel
vadorov+  4656  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-minimize
vadorov+  4655  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  4654  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicLauncher
vadorov+  4652  3332  5 Nov16 tty1     01:22:49              |   |   \_ cosmic-applet-audio
vadorov+  4608  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-input-sources
vadorov+  4651  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-battery
vadorov+  4650  3332  0 Nov16 tty1     00:00:05              |   |   \_ cosmic-applet-bluetooth
vadorov+  4653  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  4657  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-app-list
vadorov+  4596  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  4603  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-tiling
vadorov+  4605  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-power
vadorov+  4593  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  4362  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-minimize
vadorov+  4363  3332  0 Nov16 tty1     00:00:01              |   |   \_ cosmic-app-list
vadorov+  4361  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  4602  3332  0 Nov16 tty1     00:00:18              |   |   \_ cosmic-applet-time
vadorov+  4604  3332  0 Nov16 tty1     00:00:15              |   |   \_ cosmic-applet-status-area
vadorov+  4359  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  4355  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-input-sources
vadorov+  4357  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-battery
vadorov+  4356  3332  0 Nov16 tty1     00:00:05              |   |   \_ cosmic-applet-bluetooth
vadorov+  4360  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicLauncher
vadorov+  4354  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-network
vadorov+  4348  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-status-area
vadorov+  4353  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-notifications
vadorov+  4346  3332  0 Nov16 tty1     00:00:18              |   |   \_ cosmic-applet-time
vadorov+  4338  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  4341  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  4347  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-tiling
vadorov+  4349  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-power
vadorov+  4358  3332  5 Nov16 tty1     01:22:47              |   |   \_ cosmic-applet-audio
vadorov+  4606  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-notifications
vadorov+  3776  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  3771  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  3777  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-minimize
vadorov+  3751  3332  0 Nov16 tty1     00:00:05              |   |   \_ cosmic-applet-bluetooth
vadorov+  3750  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-input-sources
vadorov+  3752  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-battery
vadorov+  3706  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-notifications
vadorov+  3699  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-power
vadorov+  3709  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-network
vadorov+  3694  3332  0 Nov16 tty1     00:00:20              |   |   \_ cosmic-applet-time
vadorov+  3696  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-tiling
vadorov+  3757  3332  5 Nov16 tty1     01:23:52              |   |   \_ cosmic-applet-audio
vadorov+  3774  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicLauncher
vadorov+  3680  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicWorkspaces
vadorov+  6746  3680  0 Nov16 tty1     00:00:00              |   |   |   \_ [cosmic-workspac] <defunct>
vadorov+  6552  3680  0 Nov16 tty1     00:00:00              |   |   |   \_ [cosmic-workspac] <defunct>
vadorov+  3697  3332  0 Nov16 tty1     00:00:15              |   |   \_ cosmic-applet-status-area
vadorov+  3691  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-panel-button com.system76.CosmicAppLibrary
vadorov+  3781  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-app-list
vadorov+  4607  3332  0 Nov16 tty1     00:00:00              |   |   \_ cosmic-applet-network
vadorov+  3324  3273  0 Nov16 tty1     00:00:00              |   \_ cosmic-settings-daemon
vadorov+  3328  3324  0 Nov16 tty1     00:00:00              |   |   \_ /usr/libexec/geoclue-2.0/demos/agent
vadorov+  3337  3273  0 Nov16 tty1     00:00:00              |   \_ cosmic-app-library
vadorov+  3277  3273  7 Nov16 tty1     01:46:09              |   \_ cosmic-comp
vadorov+  3286  3277  0 Nov16 tty1     00:05:23              |   |   \_ Xwayland :1 -verbose -rootless -terminate -wm 41 -displayfd 48 -listenfd 45 -listenfd 46
vadorov+  3487  3273  0 Nov16 tty1     00:00:00              |   \_ cosmic-idle
vadorov+  3272  3270  0 Nov16 tty1     00:00:00              \_ dbus-daemon --nofork --print-address 4 --session
```

**Reply by WhyNotHugo** [reply]
If you start D-Bus using `dbus-daemon --nofork --print-address 4 --session`, it will use a random path for its socket. E.g.:

    unix:path=/tmp/dbus-XKkZ9gN05f,guid=1a6caf814103df7046891dea675c2dc4

Clients will then be unable to find the D-Bus socket. There are two way that you can work around this:

1. Use `dbus-daemon --session --address unix:path=${XDG_RUNTIME_DIR}/bus ...`. This path is the default path, and most clients will attempt to use this path if `$DBUS_SESSION_BUS_ADDRESS` is unset. [The sole exception is the godbus](https://github.com/godbus/dbus/pull/385) (and programs that use it).
2. When D-Bus prints its address to fd=4, read this data, and export it as an environment variable to all processes that could potentially rely on D-Bus.

I recommend using the first approach: hard-code a path where you want the socket. You can also `
export DBUS_SESSION_BUS_ADDRESS="unix:path=$XDG_RUNTIME_DIR/bus"` for your entire process tree so that even exceptional programs find the path.

---

FYI: On `systemd` setups, a pam module will auto-start D-Bus when you log in, and export `$DBUS_SESSION_BUS_ADDRESS` for your entire user session.

**Reply by WhyNotHugo** [reply]
I now see that you're using `/usr/bin/dbus-run-session` which (I think) does exactly what I described for you.

**Comment by simonvanderveldt** [comment]
@vadorovsky Thanks for trying this out and providing some feedback!

I'm having to do the same thing as you, there's no autostart support in Cosmic yet, the stuff that gets started for a lot of people automatically is actually started by systemd via xdg-desktop-autostart.target and a generator https://www.freedesktop.org/software/systemd/man/latest/systemd-xdg-autostart-generator.html
If you have enough rust experience to create a PR for autostart support in cosmic-session that would be great. Hope they will accept it :crossed_fingers: 

You mentioned not wanting to use GDM, that was in the context of maybe having to switch to it for autostart support?
I see in your process tree you're using greetd, is that working well in general? I was considering switching to that as well.
I looked into if there was a simple way to (auto)start some processes on login via greetd, but couldn't really find any. The main thing I've found was to use something like https://github.com/chimera-linux/turnstile together with dinit (https://github.com/davmac314/dinit/blob/master/doc/getting_started.md) to automatically start your (dinit) user services on login/user session creation. Was a bit too much to figure out quickly, so I parked that for later for now.

Btw your process tree is showing applets multiple times, do you have any idea why this is?

**Reply by vadorovsky** [reply]
> @vadorovsky Thanks for trying this out and providing some feedback!
> 
> I'm having to do the same thing as you, there's no autostart support in Cosmic yet, the stuff that gets started for a lot of people automatically is actually started by systemd via xdg-desktop-autostart.target and a generator https://www.freedesktop.org/software/systemd/man/latest/systemd-xdg-autostart-generator.html If you have enough rust experience to create a PR for autostart support in cosmic-session that would be great. Hope they will accept it 🤞

I started working on the autostart change. I'm hoping to have it ready in a week or so. :crossed_fingers: 

BTW, they merged my small fix for screen sharing. https://github.com/pop-os/xdg-desktop-portal-cosmic/pull/109 After applying this patch and after running `/usr/bin/gentoo-pipewire-launcher restart` manually, my screen sharing works without any issues.

> You mentioned not wanting to use GDM, that was in the context of maybe having to switch to it for autostart support? I see in your process tree you're using greetd, is that working well in general? I was considering switching to that as well.

In general, it's working well. I have no issues so far.

>I looked into if there was a simple way to (auto)start some processes on login via greetd, but couldn't really find any. The main thing I've found was to use something like https://github.com/chimera-linux/turnstile together with dinit (https://github.com/davmac314/dinit/blob/master/doc/getting_started.md) to automatically start your (dinit) user services on login/user session creation. Was a bit too much to figure out quickly, so I parked that for later for now.

Interesting find with turnstile. And yes, I also don't think there is any autostart feature in greetd. Other wayland compositors like sway or hyprland usually let you define scripts where you can specify commands to autostart.

I think XDG autostart files and launching the commands specified by them as children of cosmic-session would be good immediate solution. In my PR, I'm planning to introduce a CLI argument to `cosmic-session` for it.

For the long-term solution for Gentoo, I believe that OpenRC support for user services would be the best.

> Btw your process tree is showing applets multiple times, do you have any idea why this is?

Yes, that's because I'm using 3 displays and I think each applet instance on each display is a separate process. :sweat_smile:

**Reply by simonvanderveldt** [reply]
> I started working on the autostart change. I'm hoping to have it ready in a week or so. 🤞
> 
> BTW, they merged my small fix for screen sharing. [pop-os/xdg-desktop-portal-cosmic#109](https://github.com/pop-os/xdg-desktop-portal-cosmic/pull/109) After applying this patch and after running `/usr/bin/gentoo-pipewire-launcher restart` manually, my screen sharing works without any issues.

Nice to see that PRs from non-core contributors are being merged, hope all goes well with the autostart PR :crossed_fingers: 



> In general, it's working well. I have no issues so far.



> Interesting find with turnstile. And yes, I also don't think there is any autostart feature in greetd. Other wayland compositors like sway or hyprland usually let you define scripts where you can specify commands to autostart.
> 
> I think XDG autostart files and launching the commands specified by them as children of cosmic-session would be good immediate solution. In my PR, I'm planning to introduce a CLI argument to `cosmic-session` for it.
> 
> For the long-term solution for Gentoo, I believe that OpenRC support for user services would be the best.

I'm probably going to try greetd + turnstile + dinit, curiout to try it out anyway :)
If you have a autostart PR for cosmic-session feel free to ping me, I can try it out and give some feedback in the PR as well.



> Yes, that's because I'm using 3 displays and I think each applet instance on each display is a separate process. 😅

Ah, that makes sense, didn't consider that option :)

**Comment by vadorovsky** [comment]
I fixed the Steam issue here:

https://github.com/pop-os/cosmic-comp/pull/1038

I'm still working on the autostart stuff, but hoping to wrap it up soon.

**Comment by mainrs** [comment]
I've read on a GitHub issue that `.config/autostart` allegedly works. It didn't on my system. My guess is that they didn't use COSMIC's login manager.

I also have issues with pipewire on the system. I start it automatically as a user service, but that is "too late": the panel's volume applet already launched before the piprewire server started. Therefore, it does not contain any volume sliders and the like.

As of now, I'd only recommend the COSMIC alpha on systemd.

**Reply by vadorovsky** [reply]
Can you link the issue? My understanding is that `.config/autostart` works only on systemd.

**Comment by fsvm88** [comment]
Hey guys, sorry for dropping a bit out of the discussion, I was waiting for things to mature a bit, [and they did](https://github.com/pop-os/cosmic-session/pull/109) (sorry for missing this, March and April were really bad for me time-wise).

I opened PR #61 to enable `--features autostart` for the non-systemd `src_configure` path.

Can you test it and report?

**Comment by zChris83** [comment]
Im currently testing out COSMIC on laptop with primus and OpenRC. It seems to me that system76-power doesnt work correctly. I hacve to start it manually with system76-power daemon, then when i try to change to hybrid graphics (becouse cosmic-comp is running on dGPU) it gives me this  error:

[INFO] Creating /etc/modprobe.d/system76-power.conf
org.freedesktop.DBus.Error.Failed: failed to execute systemctl command: No such file or directory (os error 2)

Im also using SDDM as a login manager.
